modules/renter/proto/renew.go

```diff
@@ -0,0 +1,247 @@
+	// extract vars from params, for convenience
+	host, filesize, startHeight, endHeight, refundAddress := params.Host, params.Filesize, params.StartHeight, params.EndHeight, params.RefundAddress
+	ourSK := contract.SecretKey
+
+	// calculate cost to renter and cost to host
+	storageAllocation := host.StoragePrice.Mul64(filesize).Mul64(uint64(endHeight - startHeight))
+	hostCollateral := host.Collateral.Mul64(filesize).Mul64(uint64(endHeight - startHeight))
+	if hostCollateral.Cmp(host.MaxCollateral) > 0 {
+		// TODO: if we have to cap the collateral, it probably means we shouldn't be using this host
+		// (ok within a factor of 2)
+		hostCollateral = host.MaxCollateral
+	}
+
+	timeExtension := uint64((endHeight + host.WindowSize) - contract.LastRevision.NewWindowEnd)
+	basePrice := host.StoragePrice.Mul64(contract.LastRevision.NewFileSize).Mul64(timeExtension)    // cost of data already covered by contract, i.e. lastrevision.Filesize
+	baseCollateral := host.Collateral.Mul64(contract.LastRevision.NewFileSize).Mul64(timeExtension) // same but collateral
+
+	hostPayout := hostCollateral.Add(host.ContractPrice).Add(basePrice)
+
+	payout := storageAllocation.Add(hostCollateral.Add(host.ContractPrice)).Mul64(10406).Div64(10000) // renter covers siafund fee
+	renterCost := payout.Sub(hostCollateral)
+
+	// create file contract
+	fc := types.FileContract{
+		FileSize:       contract.LastRevision.NewFileSize,
+		FileMerkleRoot: contract.LastRevision.NewFileMerkleRoot,
+		WindowStart:    endHeight,
+		WindowEnd:      endHeight + host.WindowSize,
+		Payout:         payout,
+		UnlockHash:     contract.LastRevision.NewUnlockHash,
+		RevisionNumber: 0,
+		ValidProofOutputs: []types.SiacoinOutput{
+			// renter
+			{Value: types.PostTax(startHeight, payout).Sub(hostPayout), UnlockHash: refundAddress},
+			// host
+			{Value: hostPayout, UnlockHash: host.UnlockHash},
+		},
+		MissedProofOutputs: []types.SiacoinOutput{
+			// renter
+			{Value: types.PostTax(startHeight, payout).Sub(hostPayout), UnlockHash: refundAddress},
+			// host gets its unused collateral back, plus the contract price
+			{Value: hostCollateral.Sub(baseCollateral).Add(host.ContractPrice), UnlockHash: host.UnlockHash},
+			// void gets the spent storage fees, plus the collateral being risked
+			{Value: basePrice.Add(baseCollateral), UnlockHash: types.UnlockHash{}},
+		},
+	}
+
+	// calculate transaction fee
+	_, maxFee := tpool.FeeEstimation()
+	fee := maxFee.Mul64(estTxnSize)
+
+	// build transaction containing fc
+	err := txnBuilder.FundSiacoins(renterCost.Add(fee))
+	if err != nil {
+		return Contract{}, err
+	}
+	txnBuilder.AddFileContract(fc)
+
+	// add miner fee
+	txnBuilder.AddMinerFee(fee)
+
+	// create initial transaction set
+	txn, parentTxns := txnBuilder.View()
+	txnSet := append(parentTxns, txn)
+
+	// initiate connection
+	conn, err := net.DialTimeout("tcp", string(host.NetAddress), 15*time.Second)
```

Comment:
The hostdb scan times out in 5 seconds, so it's a moot point. We can extend it across the board, but I think it's out of scope for this PR.

We may want to add special accommodations for Tor users. Long timeouts degrade the average user experience.